```
======================================================================================================================= FAILURES =======================================================================================================================
________________________________________________________________________________________________________________ test_shadow_ioeventfd _________________________________________________________________________________________________________________

    def test_shadow_ioeventfd():
        """Configure a shadow ioeventfd, have the client trigger it, confirm that
        the server receives the notification and can see the value."""
    
        # server setup
        ctx = vfu_create_ctx(flags=LIBVFIO_USER_FLAG_ATTACH_NB)
        assert ctx is not None
        ret = vfu_setup_region(ctx, index=VFU_PCI_DEV_BAR0_REGION_IDX, size=0x1000,
                               flags=VFU_REGION_FLAG_RW)
        assert ret == 0
        fo = tempfile.TemporaryFile(dir="/dev/shm")
        fo.truncate(0x1000)
    
        # FIXME
        # Use pip install eventfd?
        #   $ grep EFD_NONBLOCK -wr /usr/include/
        #   /usr/include/bits/eventfd.h:    EFD_NONBLOCK = 00004000
        EFD_NONBLOCK = 0o00004000
    
        efd = eventfd(flags=EFD_NONBLOCK)
        ret = vfu_create_ioeventfd(ctx, VFU_PCI_DEV_BAR0_REGION_IDX, efd, 0x8,
                                   0x16, 0, 0, fo.fileno(), 0x10)
>       assert ret == 0
E       assert -1 == 0

test_shadow_ioeventfd.py:86: AssertionError
----------------------------------------------------------------------------------------------------------------- Captured stdout call -----------------------------------------------------------------------------------------------------------------
DEBUG: shadow ioeventfd not compiled
=============================================================================================================== short test summary info ================================================================================================================
FAILED test_shadow_ioeventfd.py::test_shadow_ioeventfd - assert -1 == 0
============================================================================================================ 1 failed, 223 passed in 4.32s =============================================================================================================
==62607== 
==62607== HEAP SUMMARY:
==62607==     in use at exit: 732,459 bytes in 352 blocks
==62607==   total heap usage: 49,528 allocs, 49,176 frees, 179,602,701 bytes allocated
==62607== 
==62607== 15,166 (14,624 direct, 542 indirect) bytes in 1 blocks are definitely lost in loss record 91 of 96
==62607==    at 0x488C0AC: calloc (in /usr/libexec/valgrind/vgpreload_memcheck-arm64-linux.so)
==62607==    by 0x7BAC6E7: vfu_create_ctx (libvfio-user.c:1905)
==62607==    by 0x7B16623: ??? (in /usr/lib/aarch64-linux-gnu/libffi.so.8.1.4)
==62607==    by 0x7B1383B: ??? (in /usr/lib/aarch64-linux-gnu/libffi.so.8.1.4)
==62607==    by 0x7AE2AB7: ??? (in /usr/lib/python3.12/lib-dynload/_ctypes.cpython-312-aarch64-linux-gnu.so)
==62607==    by 0x7ADF863: ??? (in /usr/lib/python3.12/lib-dynload/_ctypes.cpython-312-aarch64-linux-gnu.so)
==62607==    by 0x4C3F37: _PyObject_MakeTpCall (in /usr/bin/python3.12)
==62607==    by 0x564803: _PyEval_EvalFrameDefault (in /usr/bin/python3.12)
==62607==    by 0x4C5B03: _PyObject_Call_Prepend (in /usr/bin/python3.12)
==62607==    by 0x529DFF: ??? (in /usr/bin/python3.12)
==62607==    by 0x4C3FEF: _PyObject_MakeTpCall (in /usr/bin/python3.12)
==62607==    by 0x564803: _PyEval_EvalFrameDefault (in /usr/bin/python3.12)
==62607== 
==62607== LEAK SUMMARY:
==62607==    definitely lost: 14,624 bytes in 1 blocks
==62607==    indirectly lost: 542 bytes in 3 blocks
==62607==      possibly lost: 0 bytes in 0 blocks
==62607==    still reachable: 717,293 bytes in 348 blocks
==62607==         suppressed: 0 bytes in 0 blocks
==62607== Reachable blocks (those to which a pointer was found) are not shown.
==62607== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==62607== 
==62607== For lists of detected and suppressed errors, rerun with: -s
==62607== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
```
One test case failed, but it appears unrelated to my changes.

--- a/lib/dma.c
+++ b/lib/dma.c
@@ -66,6 +66,32 @@ fd_get_blocksize(int fd)
     return st.st_blksize;
 }
 
+static int
+dirty_page_logging_start_on_region(dma_memory_region_t *region, size_t pgsize)
+{
+    assert(region->fd != -1);
+
+    ssize_t size = get_bitmap_size(region->info.iova.iov_len, pgsize);
+    if (size < 0) {
+        return size;
+    }
+
+    region->dirty_bitmap = calloc(size, 1);
+    if (region->dirty_bitmap == NULL) {
+        return ERROR_INT(errno);
+    }
+    return 0;
+}
+
+static void
+dirty_page_logging_stop_on_region(dma_memory_region_t *region)
+{
+    if (region->dirty_bitmap != NULL) {
+        free(region->dirty_bitmap);
+        region->dirty_bitmap = NULL;
+    }
+}
+
 dma_controller_t *
 dma_controller_create(vfu_ctx_t *vfu_ctx, size_t max_regions, size_t max_size)
 {
@@ -134,6 +160,8 @@ MOCK_DEFINE(dma_controller_unmap_region)(dma_controller_t *dma,
     }
 
     assert(region->fd != -1);
+
+    dirty_page_logging_stop_on_region(region);
 
     err = fd_cache_put(&region->fd);
     assert(err == 0);
@@ -255,23 +283,6 @@ dma_map_region(dma_controller_t *dma, dma_memory_region_t *region)
             region->info.mapping.iov_base, iov_end(&region->info.mapping));
 
 
-    return 0;
-}
-
-static int
-dirty_page_logging_start_on_region(dma_memory_region_t *region, size_t pgsize)
-{
-    assert(region->fd != -1);
-
-    ssize_t size = get_bitmap_size(region->info.iova.iov_len, pgsize);
-    if (size < 0) {
-        return size;
-    }
-
-    region->dirty_bitmap = calloc(size, 1);
-    if (region->dirty_bitmap == NULL) {
-        return ERROR_INT(errno);
-    }
     return 0;
 }
 
@@ -413,7 +424,7 @@ rollback:
         if (region->info.vaddr != NULL) {
             dma_controller_unmap_region(dma, region);
         }
-        free(region->dirty_bitmap);
+        dirty_page_logging_stop_on_region(region);
         free(region);
     }
     err = fd_cache_put(&fd);
@@ -475,8 +486,7 @@ dma_controller_dirty_page_logging_reset(dma_controller_t *dma)
     for (btree_iter_init(&dma->regions, 0, &iter);
          (region = btree_iter_get(&iter, NULL)) != NULL;
          btree_iter_next(&iter)) {
-        free(region->dirty_bitmap);
-        region->dirty_bitmap = NULL;
+        dirty_page_logging_stop_on_region(region);
     }
     dma->dirty_pgsize = 0;
 }

--- a/test/py/test_dirty_pages.py
+++ b/test/py/test_dirty_pages.py
@@ -412,4 +412,90 @@ def test_dirty_pages_uninitialised_dma():
 
     vfu_destroy_ctx(ctx)
 
+
+def test_dirty_pages_ctx_no_stop():
+    """
+    Test that dirty page tracking resources are properly cleaned up even when:
+    1. DMA regions are unmapped while dirty page logging is active
+    2. The device context is destroyed without explicitly stopping dirty page
+       logging first
+
+    This validates that there are no memory leaks in the dirty bitmap handling
+    paths during both explicit DMA unmap operations and implicit cleanup during
+    context destruction.
+    """
+    global ctx, client
+
+    # Create context and initialize device
+    ctx = vfu_create_ctx(flags=LIBVFIO_USER_FLAG_ATTACH_NB)
+    assert ctx is not None
+
+    ret = vfu_pci_init(ctx)
+    assert ret == 0
+
+    vfu_setup_device_quiesce_cb(ctx, quiesce_cb=quiesce_cb)
+
+    ret = vfu_setup_device_dma(ctx, MAX_DMA_REGIONS,
+                               dma_register,
+                               dma_unregister)
+    assert ret == 0
+
+    ret = vfu_realize_ctx(ctx)
+    assert ret == 0
+
+    client = connect_client(ctx)
+
+    # Create two separate DMA regions
+    f1 = tempfile.TemporaryFile()
+    f1.truncate(0x10 << PAGE_SHIFT)
+    f2 = tempfile.TemporaryFile()
+    f2.truncate(0x10 << PAGE_SHIFT)
+
+    # Map first region at 0x100000
+    payload = vfio_user_dma_map(argsz=len(vfio_user_dma_map()),
+        flags=(VFIO_USER_F_DMA_REGION_READ | VFIO_USER_F_DMA_REGION_WRITE),
+        offset=0, addr=0x10 << PAGE_SHIFT, size=0x10 << PAGE_SHIFT)
+    msg(ctx, client.sock, VFIO_USER_DMA_MAP, payload, fds=[f1.fileno()])
+
+    # Map second region at 0x300000
+    payload = vfio_user_dma_map(argsz=len(vfio_user_dma_map()),
+        flags=(VFIO_USER_F_DMA_REGION_READ | VFIO_USER_F_DMA_REGION_WRITE),
+        offset=0, addr=0x30 << PAGE_SHIFT, size=0x10 << PAGE_SHIFT)
+    msg(ctx, client.sock, VFIO_USER_DMA_MAP, payload, fds=[f2.fileno()])
+
+    # Setup migration support
+    ret = vfu_setup_device_migration_callbacks(ctx)
+    assert ret == 0
+
+    # Start dirty page logging
+    start_logging()
+
+    # Modify pages in both regions
+    write_to_page(ctx, 0x10, 1, get_bitmap=False)
+    write_to_page(ctx, 0x30, 1, get_bitmap=False)
+
+    # Unmap the first DMA region
+    payload = vfio_user_dma_unmap(argsz=len(vfio_user_dma_unmap()),
+        addr=0x10 << PAGE_SHIFT, size=0x10 << PAGE_SHIFT)
+    msg(ctx, client.sock, VFIO_USER_DMA_UNMAP, payload)
+
+    # Verify first region is actually unmapped (should fail)
+    get_dirty_page_bitmap(addr=0x10 << PAGE_SHIFT,
+                          length=0x10 << PAGE_SHIFT,
+                          expect=errno.ENOENT)
+
+    # Verify second region still works and dirty bit is correctly set
+    bitmap = get_dirty_page_bitmap(addr=0x30 << PAGE_SHIFT,
+                                   length=0x10 << PAGE_SHIFT)
+    # First page of second region should be marked dirty
+    assert bitmap == 0b1
+
+    # Don't unmap the second DMA region - let context destruction clean it up
+    # This tests that the destroy path properly frees remaining dirty bitmaps
+
+    # Cleanup
+    client.disconnect(ctx)
+    vfu_destroy_ctx(ctx)
+
+
 # ex: set tabstop=4 shiftwidth=4 softtabstop=4 expandtab:

